### ISSUE TITLE
tokensOfOwner function will eventually gas out

### ISSUE THREAD

**Issue by ricky-dunn** [EtherBots/ETHERBOTS-Bug-Bounty#4]
Currently while there's a high chance that the below function will only ever get called off chain, in it's currently implementation of using a for loop over the entirety of `totalSupply()`, it will quickly become inefficient. 

```
function tokensOfOwner(address _owner) external view returns(uint256[] ownerTokens) {
    uint256 tokenCount = balanceOf(_owner);

    uint256[] memory result = new uint256[](tokenCount);

    uint256 totalParts = totalSupply();
    uint256 resultIndex = 0;

    for (uint partId = 0; partId < totalParts; partId++) {
        if (partIndexToOwner[partId] == _owner) {
            result[resultIndex] = partId;
            resultIndex++;
        }
    }
    return result;
}
```

While this may not pose a problem for any of your future functions it could be problematic down the line for wallets / 3rd parties who may want to show all users assets and may rely on that function to get this data out.

**Fix**
The DharmaProtocol implementation of this uses mappings and has a good example which has been fully implemented:

https://github.com/dharmaprotocol/NonFungibleToken/blob/master/contracts/NonFungibleToken.sol

Essentially they include 2 additionally mappings to keep track of tokens

```
mapping(address => uint[]) internal ownerToTokensOwned;
mapping(uint => uint) internal tokenIdToOwnerArrayIndex; 
```

And then they do a hot swap and length reduction when a token is removed.

**Comment by EtherBots** [comment]
This was intentional to minimise storage costs. We'll look further into it but it's definitely not a bug ATM. Thanks!

**Comment by ricky-dunn** [comment]
Noted. I guess just be aware that at some point it will become a bug in that it just won't return, one of CK's guys mentioned that their similar implementation stopped working completely on Geth after 100k tokens. 

> For example, our implementation of tokensOfOwner() has been entirely non-functional on geth (and I assume other full nodes) ever since we crossed 100k NFTs or so.

From https://github.com/ethereum/eips/issues/721

**Comment by EtherBots** [comment]
@ricky-dunn Ah that is very interesting (and if we change we will indeed award this bounty!) Thank you.

**Comment by RobbieFerguson** [comment]
@ricky-dunn Just looking at this -- is it a problem if it is never called by a contract? We only intend to use it for reading state.

**Comment by DanielRX** [comment]
@ricky-dunn @RobbieFerguson it's explicitly listed as not to be called by a contract in any other code I have see, as it is gas hungry. It's a web3 utility function, NOT a contract to contract call. I don't see a use contract -> contract for it.

**Comment by RobbieFerguson** [comment]
@DanielMReed Yep, we have the same understanding and would never use it an inter-contract call.
My question is, will the web3 view function fail given a big enough data set to loop over? (For reasons unrelated to gas). Still confused about this. Thanks!

**Comment by ricky-dunn** [comment]
@RobbieFerguson - even off-chain this will eventually prove problematic once a certain number of tokens is reached given it's a straight up loop over the entire supply.

In CryptoKitties case it looks like after reaching 100k tokens it no longer became functional off-chain via a full node, although not sure if they had tried to access this using a dedicated node or an Infura shared node, I suspect that may vary your mileage.

I appreciate as it's not used on-chain there's probably not a huge amount of focus, but if you did want to make if functional but not incur the overhead of storing part indexes on a per user basis (which would be a fair challenge given storage costs) you could consider including a function where you can pass in a `startIndex` and `limit` so off-chain callers can at least paginate their result sets, although this may be susceptible to race conditions so there is a trade-off there, but that's quite often the case with most things on the blockchain.

**Comment by RobbieFerguson** [comment]
@ricky-dunn We've made changes to the code based on this issue and so will discuss an appropriate bounty! Thanks.
